crates/sui/src/client_ptb/error.rs

```diff
@@ -0,0 +1,189 @@
+            span: $l,
+            help: None,
+        }
+    };
+    ($l:expr => help: { $($h:expr),* }, $($arg:tt)*) => {
+        $crate::client_ptb::error::PTBError {
+            message: format!($($arg)*),
+            span: $l,
+            help: Some(format!($($h),*)),
+        }
+    };
+}
+
+pub use sp;
+
+impl PTBError {
+    /// Add a help message to an error.
+    pub fn with_help(self, help: String) -> Self {
+        let PTBError {
+            message,
+            span,
+            help: _,
+        } = self;
+        PTBError {
+            message,
+            span,
+            help: Some(help),
+        }
+    }
+}
+
+impl Span {
+    /// Wrap a value with a span.
+    pub fn wrap<T>(self, value: T) -> Spanned<T> {
+        Spanned { span: self, value }
+    }
+
+    /// Widen the span to include another span. The resulting span will start at the minimum of the
+    /// two start positions and end at the maximum of the two end positions.
+    pub fn widen(self, other: Span) -> Span {
+        Span {
+            start: self.start.min(other.start),
+            end: self.end.max(other.end),
+        }
+    }
+
+    /// Widen the span to include another if it is Some, otherwise return the original span.
+    pub fn widen_opt(self, other: Option<Span>) -> Span {
+        other.map_or(self, |other| self.widen(other))
+    }
+
+    /// Create a span that points to the end of the file/string contents.
+    pub fn eof_span() -> Span {
+        Self {
+            start: usize::MAX,
+            end: usize::MAX,
+        }
+    }
+}
+
+impl<T> Spanned<T> {
+    /// Apply a function `f` to the underlying value, returning a new `Spanned` with the same span.
+    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
+        Spanned {
+            span: self.span,
+            value: f(self.value),
+        }
+    }
+
+    /// Widen the span to include another span. The resulting span will start at the minimum of the
+    /// two start positions and end at the maximum of the two end positions.
+    pub fn widen<U>(self, other: Spanned<U>) -> Spanned<T> {
+        self.widen_span(other.span)
+    }
+
+    /// Widen the span to include another span. The resulting span will start at the minimum of the
+    /// two start positions and end at the maximum of the two end positions.
+    pub fn widen_span(self, other: Span) -> Spanned<T> {
+        Spanned {
+            span: self.span.widen(other),
+            value: self.value,
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Spanned<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Spanned")
+            .field("span", &self.span)
+            .field("value", &self.value)
+            .finish()
+    }
+}
+
+impl<T: Clone> Clone for Spanned<T> {
+    fn clone(&self) -> Self {
+        Spanned {
+            span: self.span,
+            value: self.value.clone(),
+        }
+    }
+}
+
+impl<T: Copy> Copy for Spanned<T> {}
```

Comment:
I don't understand this, is it just a sort of `#[derive(Copy)]` for Spanned?

Reply:
Kinda, it's more of a conditional derivation of `Copy` for `Spanned`. Basically "if you give me a `Copy` `T`, then you can copy `Spanned<T>`. With a `derive(Copy)` on `Spanned` it would require that all `T`'s that it held would have to be `Copy`, whereas this you can have `Spanned<T>`s where `T` cannot be copied (and `Spanned<T>` can't), but if `T` can be copied then `Spanned<T>` can.

Reply:
Got it, thanks!